### PR TITLE
Add functions to enable/disable audio button

### PR DESCRIPTION
**Pull request by otacke**
Some content types may require to disable/enable the audio button conditionally. This pull request will add the required functionality.

Disabling is not done using the disabled property as it implicitly sets tabindex to -1 and seems to confuse ChromeVox navigation when the button that's disabled is in focus.

### CODE REVIEW

File: scripts/audio.js [hunk 1]
```diff
@@ -61,9 +61,14 @@ H5P.Audio = (function ($) {
       'class': AUDIO_BUTTON + " " + PLAY_BUTTON
     }).appendTo(self.$inner)
       .click( function () {
+        if (!self.isEnabled) {
```
Comment (fnoks) [inline scripts/audio.js]: self.isEnabled is not defined. I guess you meant self.isEnabled() ?

File: scripts/audio.js [hunk 2]
```diff
@@ -94,6 +99,14 @@ H5P.Audio = (function ($) {
       audioButton.removeClass(PAUSE_BUTTON).addClass(PLAY_BUTTON_PAUSED);
     });
 
+    // Prevent play/pause interaction on button when disabled.
+    audioButton.get(0).addEventListener('keydown', function (event) {
+      event = event || window.event;
+      if (!self.isEnabled() && (event.keyCode === 13 || event.keyCode === 32)) {
+        event.preventDefault();
+      }
+    }, false);
```
Comment (fnoks) [inline scripts/audio.js]: I don't understand why we need this? This keydown results in a click event, right? And in the click event handler we are already checking for isEnabled...

File: scripts/audio.js [hunk 3]
```diff
@@ -317,3 +330,42 @@ H5P.Audio.prototype.getCurrentState = function () {
...
+ * Not using disabled attribute to block button activation, because it will
+ * implicitly set tabindex = -1 and confuse ChromeVox navigation. Clicks handled
+ * using "pointer-events: none" in CSS, keydown on enter or space prevented by
+ * interception in event listener on button.
+ */
+H5P.Audio.prototype.disable = function () {
+  if (this.$audioButton) {
+    this.$audioButton.toggleClass(H5P.Audio.BUTTON_DISABLED, true);
+  }
+};
```
Comment (fnoks) [inline scripts/audio.js]: I think disable/enable could be misunderstood. I propose renaming to disableToggleButton()/enableToggleBurtton() or something better :)

File: scripts/audio.js [hunk 4]
```diff
@@ -317,3 +330,42 @@ H5P.Audio.prototype.getCurrentState = function () {
...
+/**
+ * @public
+ * Check if button is enabled.
+ * @return {boolean} True, if button is enabled. Else false.
+ */
+H5P.Audio.prototype.isEnabled = function () {
+  if (!this.$audioButton) {
+    return false;
+  }
+  return !this.$audioButton.hasClass(H5P.Audio.BUTTON_DISABLED);
```
Comment (fnoks) [inline scripts/audio.js]: We should not use class names for this, when we can use local variables instead. So, having a self.toggleButtonEnabled (which is default set to true) is much better.

File: scripts/audio.js [hunk 5]
```diff
@@ -317,3 +330,42 @@ H5P.Audio.prototype.getCurrentState = function () {
...
+    this.$audioButton.toggleClass(H5P.Audio.BUTTON_DISABLED, false);
+  }
+};
+
+/**
+ * @public
+ * Check if button is enabled.
+ * @return {boolean} True, if button is enabled. Else false.
+ */
+H5P.Audio.prototype.isEnabled = function () {
```
Comment (fnoks) [inline scripts/audio.js]: Same thing applies here for the naming. isEnabled is too generic.

File: scripts/audio.js [hunk 6]
```diff
@@ -317,3 +330,42 @@ H5P.Audio.prototype.getCurrentState = function () {
...
+ * @public
+ * Disable button.
+ * Not using disabled attribute to block button activation, because it will
+ * implicitly set tabindex = -1 and confuse ChromeVox navigation. Clicks handled
+ * using "pointer-events: none" in CSS, keydown on enter or space prevented by
+ * interception in event listener on button.
+ */
+H5P.Audio.prototype.disable = function () {
+  if (this.$audioButton) {
+    this.$audioButton.toggleClass(H5P.Audio.BUTTON_DISABLED, true);
```
Comment (fnoks) [inline scripts/audio.js]: I would prefer addClass instead of toggleClass...

File: scripts/audio.js [hunk 7]
```diff
@@ -317,3 +330,42 @@ H5P.Audio.prototype.getCurrentState = function () {
...
+  }
+};
+
+/**
+ * @public
+ * Enable button.
+ */
+H5P.Audio.prototype.enable = function () {
+  if (this.$audioButton) {
+    this.$audioButton.toggleClass(H5P.Audio.BUTTON_DISABLED, false);
```
Comment (fnoks) [inline scripts/audio.js]: I would prefer removeClass instead of toggleClass...

File: scripts/audio.js (hunk 1 above)
Comment (otacke) [reply scripts/audio.js]: Yup, sorry I missed that.

File: scripts/audio.js (hunk 2 above)
Comment (otacke) [reply scripts/audio.js]: You're right, redundant and removed.

File: scripts/audio.js (hunk 3 above)
Comment (otacke) [reply scripts/audio.js]: changed, also changes isEnabled to isEnabledToggleButton to prevent future confusion

File: scripts/audio.js (hunk 4 above)
Comment (otacke) [reply scripts/audio.js]: Changed

File: scripts/audio.js (hunk 5 above)
Comment (otacke) [reply scripts/audio.js]: Changed

File: scripts/audio.js (hunk 6 above)
Comment (otacke) [reply scripts/audio.js]: Changed

File: scripts/audio.js (hunk 7 above)
Comment (otacke) [reply scripts/audio.js]: Changed

**Comment by otacke** [comment]
@fnoks I changed the issues. And no, I've had a look at the full player, and I could probably create some trickery with additional listeners to prevent play/pause, but I am not sure that there's a way to visually indicate that the button is disabled on all browsers using the audio element. To do this, I assume that custom controls are necessary.

**Comment by fnoks** [comment]
Thank you for your contribution, @otacke :1st_place_medal:

**Comment by otacke** [comment]
@fnoks You're more then welcome! When is release day? I need to prepare a new box of candy ... :-D

**Comment by fnoks** [comment]
Hehe, I haven't set a release date. But the box of candy will definitively make it get a higher priority :chocolate_bar:

**Pull request merged**